graylog2-server/src/main/java/org/graylog/plugins/sidecar/rest/resources/CollectorResource.java

```diff
@@ -236,31 +257,94 @@ public Response deleteCollector(@ApiParam(name = "id", required = true)
+    public ValidationResult validateCollector(
+            @Valid @ApiParam("collector") Collector toValidate) {
+        return validate(toValidate);
     }
 
-    private ValidationResponse validate(String name, String id) {
-        if (!name.matches("^[A-Za-z0-9_.-]+$")) {
-            return ValidationResponse.create(true, "Collector name can only contain the following characters: A-Z,a-z,0-9,_,-,.");
-        }
+    private ValidationResult validate(Collector toValidate) {
+        final Optional<Collector> collectorOptional;
         final Collector collector;
-        if (id == null) {
-            collector = collectorService.findByName(name);
+        final ValidationResult validation = new ValidationResult();
+
+        if (toValidate.name().isEmpty()) {
+            validation.addError("name", "Collector name cannot be empty.");
+        } else if (!validateCollectorName(toValidate.name())) {
+                validation.addError("name", "Collector name can only contain the following characters: A-Z,a-z,0-9,_,-,.");
+        }
+
+        if (toValidate.executablePath().isEmpty()) {
+            validation.addError("executable_path", "Collector binary path cannot be empty.");
+        } else if (!validatePath(toValidate.executablePath())) {
+                validation.addError("executable_path", "Collector binary path cannot contain the following characters: ; * ? \" < > | &");
+        }
+
+        if (toValidate.defaultTemplate().isEmpty()) {
+            validation.addError("default_template", "Collector default template cannot be empty.");
+        }
+
+        if (toValidate.nodeOperatingSystem() != null) {
+            if (!validateOperatingSystem(toValidate.nodeOperatingSystem())) {
+                validation.addError("node_operating_system", "Operating system can only be 'linux' or 'windows'.");
+            }
+            if (!validateServiceType(toValidate.serviceType(), toValidate.nodeOperatingSystem())) {
+                validation.addError("service_type", "Linux collectors only support 'Foreground execution' while Windows collectors additionally support 'Windows service'.");
+            }
+            collectorOptional = Optional.ofNullable(collectorService.findByNameAndOs(toValidate.name(), toValidate.nodeOperatingSystem()));
         } else {
-            collector = collectorService.findByNameExcludeId(name, id);
+            collectorOptional = Optional.ofNullable(collectorService.findByName(toValidate.name()));
         }
-        if (collector == null) {
-            return ValidationResponse.create(false, null);
+        if (collectorOptional.isPresent()) {
+            collector = collectorOptional.get();
+            if (!collector.id().equals(toValidate.id())) {
+                // a collector exists with a different id, so the name is already in use, fail validation
+                validation.addError("name", "Collector with name \"" + toValidate.name() + "\" already exists");
```

Comment:
As the combination of name and OS is what we validate, we should mention that the name exists for that OS.